build: Fix Boost.Process check for Boost 1.73 and older

hebasto opened the pull request:
On master (5f44c5c428b696af4214b2519cb2bbeb0e4a1027) Boost.Process check false fails without the `-lpthread` flag.

```
$ grep -C 2 pthread_detach config.log 
/usr/bin/ld: /tmp/cczCQfQv.o: in function `boost::asio::detail::posix_global_impl<boost::asio::system_context>::~posix_global_impl()':
conftest.cpp:(.text._ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED2Ev[_ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED5Ev]+0xa3): undefined reference to `pthread_join'
/usr/bin/ld: conftest.cpp:(.text._ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED2Ev[_ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED5Ev]+0xc4): undefined reference to `pthread_detach'
collect2: error: ld returned 1 exit status
configure:26674: $? = 1
```

Not required for Boost 1.74+.

hebasto committed:
build, refactor: Replace tabs with spaces

hebasto committed:
build: Fix a non-portable use of `test`

hebasto committed:
build: Fix Boost.Process check for Boost 1.73 and older

hebasto committed:
ci: Force `--enable-external-signer` to prevent future regressions

Sjors commented on configure.ac:
```diff
@@ -1438,9 +1438,12 @@ if test "$use_external_signer" != "no"; then
     *)
       AC_MSG_CHECKING([whether Boost.Process can be used])
+      TEMP_LDFLAGS="$LDFLAGS"
+      LDFLAGS="$TEMP_LDFLAGS $PTHREAD_CFLAGS"
```
Maybe document why this is necessary and when it can be droped, e.g.:
```
dnl This check requires the -lpthread flag.
del This workaround can be dropped when the minimum Boost is 1.78 (possibly earlier)
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24397#issuecomment-1047679471).

Sjors reviewed (approved):
utACK 2d911f2c22d2abcf1be7b63a36133a5d88cdc962

Unfortunately this change doesn't fix the broken Boost.Process on macOS for Homebrew boost 1.78 so I'll open another issue for that.

hebasto commented:
Updated 2d911f2c22d2abcf1be7b63a36133a5d88cdc962 -> 2d0d0c58d1387fb7d3cbc7118724d623e649126e ([pr24397.02](https://github.com/hebasto/bitcoin/commits/pr24397.02) -> [pr24397.03](https://github.com/hebasto/bitcoin/commits/pr24397.03), [diff](https://github.com/hebasto/bitcoin/compare/pr24397.02..pr24397.03)):

- addressed @Sjors's [comment](https://github.com/bitcoin/bitcoin/pull/24397#discussion_r811729143)

MarcoFalke commented on ci/test/00_setup_env_arm.sh:
```diff
@@ -25,4 +25,4 @@ export RUN_FUNCTIONAL_TESTS=false
 # -Wno-psabi is to disable ABI warnings: "note: parameter passing for argument of type ... changed in GCC 7.1"
 # This could be removed once the ABI change warning does not show up by default
-export BITCOIN_CONFIG="--enable-reduce-exports CXXFLAGS=-Wno-psabi"
+export BITCOIN_CONFIG="--enable-external-signer --enable-reduce-exports CXXFLAGS=-Wno-psabi"
```
Is it possible to add this to `BITCOIN_CONFIG_ALL` and then opt out in the windows build with `--disable-external-signer`?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24397#issuecomment-1047748269).

hebasto commented:
Updated 2d0d0c58d1387fb7d3cbc7118724d623e649126e -> 774323e378acea998069980edf50e1a26f4b6b9e ([pr24397.03](https://github.com/hebasto/bitcoin/commits/pr24397.03) -> [pr24397.04](https://github.com/hebasto/bitcoin/commits/pr24397.04), [diff](https://github.com/hebasto/bitcoin/compare/pr24397.03..pr24397.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/24397#discussion_r811829607)

hebasto commented:
@MarcoFalke Why removed this from the [23.0](https://github.com/bitcoin/bitcoin/milestone/52) milestone?

Otherwise, our Guix binaries will lack external signer support.

MarcoFalke commented:
I wasn't aware this affects guix builds

hebasto commented:
> I wasn't aware this affects guix builds

It does because we use boost 1.71.0 in depends.

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
6d3bc8aea9ba022a0cb7171b9a690a9c5798e838fd7f3c73212c33066cbad3f7  guix-build-774323e378ac/output/aarch64-linux-gnu/SHA256SUMS.part
d35538cf88e4862cd7d48c048d0e78547f5a44b09aa90b2cfa797530c193d854  guix-build-774323e378ac/output/aarch64-linux-gnu/bitcoin-774323e378ac-aarch64-linux-gnu-debug.tar.gz
8887d905d2d690f4f86a40992379db80732cd54cbb01ad142e4945c7ca553690  guix-build-774323e378ac/output/aarch64-linux-gnu/bitcoin-774323e378ac-aarch64-linux-gnu.tar.gz
544271d9bc866b4525bf3b2260605043ac7e9fed43802e490e2ca0219c59d0a9  guix-build-774323e378ac/output/arm-linux-gnueabihf/SHA256SUMS.part
0329772e725275e73235087ca6d0291947f4bb1800bd70653b055829cae01b62  guix-build-774323e378ac/output/arm-linux-gnueabihf/bitcoin-774323e378ac-arm-linux-gnueabihf-debug.tar.gz
28a5fa4c9150095084b02345f745d07e97a2633b68da1dc3eb6f4c280b516787  guix-build-774323e378ac/output/arm-linux-gnueabihf/bitcoin-774323e378ac-arm-linux-gnueabihf.tar.gz
48c425c5a80d18379883d7ba5db910d1031a34430b9bfe31e9ffb595ab6bd6ed  guix-build-774323e378ac/output/arm64-apple-darwin/SHA256SUMS.part
85aee9958c3ca195b3908328c792c8aa46ab3af488923afd0b149f297e05ff93  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-arm64-apple-darwin.tar.gz
fc4c40ade7b00377f64b728f8a19cfc6939a56e6f966d3b99b6c373d9c90ecd4  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.dmg
86bb7442b012585c9751c9a937df3d77c1fb2e45e40f9962cd79f55fd0bc2aa5  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.tar.gz
f7b04de225580810c9e5f8999f2e854a36b1fbe5d7c464b2fa5339f831f9fe74  guix-build-774323e378ac/output/dist-archive/bitcoin-774323e378ac.tar.gz
1bce7fdb529f35dedb239ed2c6de1f611caeabdad309a49df26feb9b6e76d370  guix-build-774323e378ac/output/powerpc64-linux-gnu/SHA256SUMS.part
f3460faf9724470f7c3c70785c1ef3f5d344f71e4230687beab8148843c7041b  guix-build-774323e378ac/output/powerpc64-linux-gnu/bitcoin-774323e378ac-powerpc64-linux-gnu-debug.tar.gz
26b8b4bf5f160c5b5926d60346842d65d15e26aae60fec7bac17e144e3bb3f2c  guix-build-774323e378ac/output/powerpc64-linux-gnu/bitcoin-774323e378ac-powerpc64-linux-gnu.tar.gz
72ecb72767d46c26b02ddb9a541e7504d6e84137283f229132728179957a7260  guix-build-774323e378ac/output/powerpc64le-linux-gnu/SHA256SUMS.part
68ff1181a88ad9de08e0310c73f81cf3ceed2db7810ab1059eed8acbc20bac4a  guix-build-774323e378ac/output/powerpc64le-linux-gnu/bitcoin-774323e378ac-powerpc64le-linux-gnu-debug.tar.gz
e691a201d071abe1fb7122fffdf8916bd5a2b01475f937bc37d1004cf773c5ef  guix-build-774323e378ac/output/powerpc64le-linux-gnu/bitcoin-774323e378ac-powerpc64le-linux-gnu.tar.gz
8261d6b583d33f36f0e515064c6e1278054e09fa2eb779a89b25dc5f964d5f64  guix-build-774323e378ac/output/riscv64-linux-gnu/SHA256SUMS.part
230981ac50dd56d5914b726364da1bda0766addb5cc16c32c37ea9e3c41cebee  guix-build-774323e378ac/output/riscv64-linux-gnu/bitcoin-774323e378ac-riscv64-linux-gnu-debug.tar.gz
f8c634f97a08f9f359b32328146752fd00621900d14e02ea4da2d9b556e45976  guix-build-774323e378ac/output/riscv64-linux-gnu/bitcoin-774323e378ac-riscv64-linux-gnu.tar.gz
b55cf1d4a77c4f096a08ed042da5602d98e820d3b7c2bb9e332c93ba39821874  guix-build-774323e378ac/output/x86_64-apple-darwin/SHA256SUMS.part
cb64acd4272f9225ee69040d1422a1f53260942d6b63b56245b8f990eb12989f  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.dmg
d5ed3d7951a2cd0c43219d348d9c6925c86073e5ccd3d7b4e8d79db59b22154b  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.tar.gz
37b52e8188a3cef88c66e9e17a9e8dd7cf048e119470c77396d3d1bb26c5b1a3  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx64.tar.gz
ea1f4f3ff3ab9b0d6dc7c307a476d01eb77411e0cec8c2da39a3be6c507e632b  guix-build-774323e378ac/output/x86_64-linux-gnu/SHA256SUMS.part
1036fe95582b8c111f17528c047ed19e8377720d4b7291edafd6cdcb3e59c634  guix-build-774323e378ac/output/x86_64-linux-gnu/bitcoin-774323e378ac-x86_64-linux-gnu-debug.tar.gz
e1fbd135d56dd6855acbe27480ec76c3b4c548f8ae26427b8aa7d197a835c4a7  guix-build-774323e378ac/output/x86_64-linux-gnu/bitcoin-774323e378ac-x86_64-linux-gnu.tar.gz
c13bca92a47a7c04e3f8f3aef2c755bd37db91a134ddb65f6e9f0da4fb8dcd72  guix-build-774323e378ac/output/x86_64-w64-mingw32/SHA256SUMS.part
d2ae028987d096a74c06d496949987e1ef015c7ef54285a09dc98af18ab9a2b1  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win-unsigned.tar.gz
0b56bb298195ef344639d8079d28cf8bd755f911babe1823e5ff3b034486d2b6  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64-debug.zip
ddc4a904c3a7a4403a7596422a6e4c81ede31ae1ecad3e7291a20eeef7586302  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64-setup-unsigned.exe
7346baae12f4b11bd0a11267c1e3bcffd3e022ac9f0a80fb6e2e7cb5c09453f9  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64.zip
```

hebasto commented:
@MarcoFalke 
> > I wasn't aware this affects guix builds
> 
> It does because we use boost 1.71.0 in depends.

It is not the case now as #24383 has just been merged.

MarcoFalke commented:
Is this for 0.23?

hebasto commented:
> Is this for 0.23?

Both this PR and #24415 are bugfixes. So I guess yes, we should strive to get both of them in 0.23.

OTOH, users with system boost <=1.73 or >=1.78 always have an opportunity to build depends.

fanquake commented:
Guix Build:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
6d3bc8aea9ba022a0cb7171b9a690a9c5798e838fd7f3c73212c33066cbad3f7  guix-build-774323e378ac/output/aarch64-linux-gnu/SHA256SUMS.part
d35538cf88e4862cd7d48c048d0e78547f5a44b09aa90b2cfa797530c193d854  guix-build-774323e378ac/output/aarch64-linux-gnu/bitcoin-774323e378ac-aarch64-linux-gnu-debug.tar.gz
8887d905d2d690f4f86a40992379db80732cd54cbb01ad142e4945c7ca553690  guix-build-774323e378ac/output/aarch64-linux-gnu/bitcoin-774323e378ac-aarch64-linux-gnu.tar.gz
544271d9bc866b4525bf3b2260605043ac7e9fed43802e490e2ca0219c59d0a9  guix-build-774323e378ac/output/arm-linux-gnueabihf/SHA256SUMS.part
0329772e725275e73235087ca6d0291947f4bb1800bd70653b055829cae01b62  guix-build-774323e378ac/output/arm-linux-gnueabihf/bitcoin-774323e378ac-arm-linux-gnueabihf-debug.tar.gz
28a5fa4c9150095084b02345f745d07e97a2633b68da1dc3eb6f4c280b516787  guix-build-774323e378ac/output/arm-linux-gnueabihf/bitcoin-774323e378ac-arm-linux-gnueabihf.tar.gz
48c425c5a80d18379883d7ba5db910d1031a34430b9bfe31e9ffb595ab6bd6ed  guix-build-774323e378ac/output/arm64-apple-darwin/SHA256SUMS.part
85aee9958c3ca195b3908328c792c8aa46ab3af488923afd0b149f297e05ff93  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-arm64-apple-darwin.tar.gz
fc4c40ade7b00377f64b728f8a19cfc6939a56e6f966d3b99b6c373d9c90ecd4  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.dmg
86bb7442b012585c9751c9a937df3d77c1fb2e45e40f9962cd79f55fd0bc2aa5  guix-build-774323e378ac/output/arm64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.tar.gz
f7b04de225580810c9e5f8999f2e854a36b1fbe5d7c464b2fa5339f831f9fe74  guix-build-774323e378ac/output/dist-archive/bitcoin-774323e378ac.tar.gz
1bce7fdb529f35dedb239ed2c6de1f611caeabdad309a49df26feb9b6e76d370  guix-build-774323e378ac/output/powerpc64-linux-gnu/SHA256SUMS.part
f3460faf9724470f7c3c70785c1ef3f5d344f71e4230687beab8148843c7041b  guix-build-774323e378ac/output/powerpc64-linux-gnu/bitcoin-774323e378ac-powerpc64-linux-gnu-debug.tar.gz
26b8b4bf5f160c5b5926d60346842d65d15e26aae60fec7bac17e144e3bb3f2c  guix-build-774323e378ac/output/powerpc64-linux-gnu/bitcoin-774323e378ac-powerpc64-linux-gnu.tar.gz
72ecb72767d46c26b02ddb9a541e7504d6e84137283f229132728179957a7260  guix-build-774323e378ac/output/powerpc64le-linux-gnu/SHA256SUMS.part
68ff1181a88ad9de08e0310c73f81cf3ceed2db7810ab1059eed8acbc20bac4a  guix-build-774323e378ac/output/powerpc64le-linux-gnu/bitcoin-774323e378ac-powerpc64le-linux-gnu-debug.tar.gz
e691a201d071abe1fb7122fffdf8916bd5a2b01475f937bc37d1004cf773c5ef  guix-build-774323e378ac/output/powerpc64le-linux-gnu/bitcoin-774323e378ac-powerpc64le-linux-gnu.tar.gz
8261d6b583d33f36f0e515064c6e1278054e09fa2eb779a89b25dc5f964d5f64  guix-build-774323e378ac/output/riscv64-linux-gnu/SHA256SUMS.part
230981ac50dd56d5914b726364da1bda0766addb5cc16c32c37ea9e3c41cebee  guix-build-774323e378ac/output/riscv64-linux-gnu/bitcoin-774323e378ac-riscv64-linux-gnu-debug.tar.gz
f8c634f97a08f9f359b32328146752fd00621900d14e02ea4da2d9b556e45976  guix-build-774323e378ac/output/riscv64-linux-gnu/bitcoin-774323e378ac-riscv64-linux-gnu.tar.gz
b55cf1d4a77c4f096a08ed042da5602d98e820d3b7c2bb9e332c93ba39821874  guix-build-774323e378ac/output/x86_64-apple-darwin/SHA256SUMS.part
cb64acd4272f9225ee69040d1422a1f53260942d6b63b56245b8f990eb12989f  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.dmg
d5ed3d7951a2cd0c43219d348d9c6925c86073e5ccd3d7b4e8d79db59b22154b  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx-unsigned.tar.gz
37b52e8188a3cef88c66e9e17a9e8dd7cf048e119470c77396d3d1bb26c5b1a3  guix-build-774323e378ac/output/x86_64-apple-darwin/bitcoin-774323e378ac-osx64.tar.gz
ea1f4f3ff3ab9b0d6dc7c307a476d01eb77411e0cec8c2da39a3be6c507e632b  guix-build-774323e378ac/output/x86_64-linux-gnu/SHA256SUMS.part
1036fe95582b8c111f17528c047ed19e8377720d4b7291edafd6cdcb3e59c634  guix-build-774323e378ac/output/x86_64-linux-gnu/bitcoin-774323e378ac-x86_64-linux-gnu-debug.tar.gz
e1fbd135d56dd6855acbe27480ec76c3b4c548f8ae26427b8aa7d197a835c4a7  guix-build-774323e378ac/output/x86_64-linux-gnu/bitcoin-774323e378ac-x86_64-linux-gnu.tar.gz
c13bca92a47a7c04e3f8f3aef2c755bd37db91a134ddb65f6e9f0da4fb8dcd72  guix-build-774323e378ac/output/x86_64-w64-mingw32/SHA256SUMS.part
d2ae028987d096a74c06d496949987e1ef015c7ef54285a09dc98af18ab9a2b1  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win-unsigned.tar.gz
0b56bb298195ef344639d8079d28cf8bd755f911babe1823e5ff3b034486d2b6  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64-debug.zip
ddc4a904c3a7a4403a7596422a6e4c81ede31ae1ecad3e7291a20eeef7586302  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64-setup-unsigned.exe
7346baae12f4b11bd0a11267c1e3bcffd3e022ac9f0a80fb6e2e7cb5c09453f9  guix-build-774323e378ac/output/x86_64-w64-mingw32/bitcoin-774323e378ac-win64.zip
```

laanwj commented:
Code review ACK 774323e378acea998069980edf50e1a26f4b6b9e, seems fine to merge last minute for 23.0.

laanwj commented on configure.ac:
```diff
@@ -1438,11 +1438,15 @@ if test "$use_external_signer" != "no"; then
+      LDFLAGS="$TEMP_LDFLAGS"
       AC_MSG_RESULT([$have_boost_process])
-      if test "$have_boost_process" == "yes"; then
+      if test "$have_boost_process" = "yes"; then
```
good catch
